### ISSUE TITLE
Why dropping composite message formatting support also for Assert.Pass/Fail/... where there are no actual and constraint parameters?

### ISSUE THREAD

**Issue by maettu-this** [nunit/nunit.analyzers#734]
[NUnit2050](https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit2050.md) tells me to replace composite message formatting by a simple string literal (not possible where variable values need to be injected) or an interpolatable string (would be possible but adds too much noise as variable names are much longer than {0} or {1}).

I fully understand the benefits of using `CallerArgumentExpression` for `Assert.That` and `Assume.That`. But why also for...
```
Assert.Pass
Assert.Fail
Assert.Warn
Assert.Ignore
Assert.Inconclusive
```
...where there neither is a `condition` nor `actual`/`constraint` parameters?

**Comment by OsirisTerje** [comment]
@manfred-brands I am also unsure here.  

The reporter seems to hold the opinion that it should stay as strings, but I don't share that an interpolated string is more noisy than an interpolated string, rather the opposite:  

```csharp
($"This is {givenname}, my last name is {surname}")
```
versus
```csharp
("This is {0}, my last name is {1}",givenname,surname)
```

On the other side, we haven't had anyone complaining about this (ordinary strings here, interpolated otherwise), so I don't think people really notice.  

Four of these asserts are also in their own group in the docs (Special Assertions) and Warn is on its own.   The keyword Warn ( used for Warn.If andWarn.Unless use CAE.

For just these with strings, a CAE doesn't really add anything, but the fact that the output message tells where the message comes from, they don\t do that now, but all functions where we have CAE does that, including the classic asserts *since they re/route to the constraint ones.

Writing a 
```csharp
Assert.Warn("There is something strange here");
```
ends with an output like this:
![image](https://github.com/nunit/nunit.analyzers/assets/203432/612329eb-c441-4a08-bd5a-5e6d83b9104d)

versus if we use something with CAE:
```csharp
Warn.If(MyValue != 42, "There is something strange here");
```
ends up like this, which imho is better:
![image](https://github.com/nunit/nunit.analyzers/assets/203432/cc1ae735-fb2c-488d-98c3-7b78529d4fe5)

So... after this little "walk inside my thoughts", I think I would prefer it to use an interpolated string and adding CAE.   :-)

**Comment by maettu-this** [comment]
Let me add three pieces to consider:

### the "master" of the message

For `Assert.That|Assume` or `If`, NUnit is the message "master", in most cases the user doesn't need to add any additional message, the NUnit standard message is perfect.

For the others like `Assert.Fail` or `Assert.Inconclusive`, the developer is the message "master", the developer even is obliged to create the message.

### conditional vs. unconditional

The above shown `Warn.If` is conditional, same as `Assert.That`, i.e. CAE makes sense. But I have multiple cases where an `Assert.Inconclusive` or `Assert.Fail` is unconditional, because more complex test logic has already determined the condition. The message is used to easier locate the root cause of a failed test run, e.g. a non-responding device.

### composite message formatting vs. interpolated strings

> don't share that an interpolated string is more noisy than an interpolated string

Many may agree for the stated example, but I personally prefer composite message formatting especially in case of formatting numbers, e.g.:
```
Assert.Inconclusive("Test skipped because buffer size {0} is not supported by VSPD.", outputBufferSize);
Assert.Fail("Count of {0} exceeds the capabilities of this test!", someCollection.Count());
```
vs.
```
Assert.Inconclusive("Test skipped because buffer size {outputBufferSize} is not supported by VSPD.");
Assert.Fail("Count of {someCollection.Count()} exceeds the capabilities of this test!");
```
With composite message formatting...

- ...the main statement of the messages is easier to catch, e.g. "buffer size is not supported by VSPD".
- ...there is less noise within the message, the noise is moved to the far end of it. 

But maybe this is just personal taste.

**Comment by OsirisTerje** [comment]
@stevenaw  wrote the following about SpecFlow and their NUnit integration:

> Reports on the SpecFlow forums show that NUnit 4 "just works" despite Specflow being compiled against NUnit 3.13.1: https://support.specflow.org/hc/en-us/community/posts/12618091291292-NUnit-4-support  

> From what I can see, this is because they primarily integrate on the assertion side by only using Assert.Inconclusive(string) and Assert.Ignore(string), neither of which was moved out into ClassicAsserts. https://github.com/SpecFlowOSS/SpecFlow/blob/master/Plugins/TechTalk.SpecFlow.NUnit.SpecFlowPlugin/NUnitRuntimeProvider.cs#L15C13-L20

This means we should be very careful and only change this after talking with the SpecFlow guys.  We don't want to break them.

**Comment by stevenaw** [comment]
Thanks for cross-posting and sharing this here @OsirisTerje  Agreed, we should be careful with those methods here.

I'm unsure if there's a proposed direction here yet but I think we should be mindful of backwards compatibility in general here too as it sounds like we're considering changing or adding method signatures in the framework, which would be released as a _semver-minor_ version bump.

**Comment by OsirisTerje** [comment]
Agree.  Perhaps we should add a v5 branch, and merge all breaking changes in there.

**Comment by manfred-brands** [comment]
> Agree. Perhaps we should add a v5 branch, and merge all breaking changes in there.

We should first decide if these changes are worthwhile enough to break compatibility or reject them and don't do them.

**Comment by OsirisTerje** [comment]
Sure :-)

Imho, I'm leaning on keeping these as they are.  If we do so, the 2050 rule should not trigger on these, and documentation updated.  

Comments: @nunit/framework-team @nunit/analyzers-team

**Comment by mikkelbu** [comment]
I think I also lean towards keeping these as they are.

**Comment by stevenaw** [comment]
No strong opinions on my end either way.
I tend to err on the side of caution and compatibility when it comes to signature changes or additions. No change is, of course, safest of all.

**Comment by manfred-brands** [comment]
It seem we are in agreement to not change this again in NUnit which would require another major version change.

This means that the analyzer is correct that it warns for `Assert.Warn` etc. when compiled against `NUnit 3` as `NUnit 4` no longer supports any `params` overloads.

**Comment by maettu-this** [comment]
@OsirisTerje and @manfred-brands I am a bit confused now, as I understood...

> If we do so, the 2050 rule should not trigger on these, and documentation updated.

...to lean towards not breaking compatibility against NUnit 3, i.e. reintroduce the `params` overload for those `Assert` that only take a single string argument. But maybe I have misinterpreted the statements and they actually meant not breaking compatibility against NUnit 4.

But even if not reintroducing the `params` overload in NUnit, I think the documentation of [NUnit2050](https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit2050.md) should be refined. Today's...

>  interpreted as the actual and constraint expression strings

...just doesn't make sense for those `Assert` that don't have `actual` and `constraint` expressions.

**Comment by manfred-brands** [comment]
> > interpreted as the actual and constraint expression strings
> 
> ...just doesn't make sense for those `Assert` that don't have `actual` and `constraint` expressions.

Correct. The table could be move higher to refer to method that used to have `params` overloads.

The special section is only relevant for `Assert.That` with 3 `string` arguments:

```csharp
[TestCase("NUnit 4", "NUnit 3")]
public void TestMessage(string actual, string expected)
{
    Assert.That(actual, Is.EqualTo(expected), "Expected '{0}', but got: '{1}'", expected, actual);
}
```
